### PR TITLE
Donations: Accept person and person params

### PR DESCRIPTION
**Pull request by Rio517**
We need to save more than just a donors email address.  This update allows donation create to accept all standard person params.

* donations#create now accepts person params
* Also, I'm standardizing our error responses to 'error' rather than
'errors'

### CODE REVIEW

**Comment by Rio517** [comment]
Hey @turino , would you mind looking this over.  Also, one thing I've noticed is that locally my boot-up time and specs run times have jumped a lot just recently.  I've gone from 2.3s -> 7s app boot time and specs 7s -> 37s.   I need to investigate further, but have you seen anything similar?

File: app/controllers/v1/donations_controller.rb [hunk 1]
```diff
@@ -1,9 +1,10 @@
 class V1::DonationsController < V1::BaseController
+  before_action :set_person
```
Comment (turino) [inline app/controllers/v1/donations_controller.rb]: I would avoid this pattern (following thoughtbot's advice). Instead you could just create a method and call it in the create action (`person = create_person` or something like that)

File: spec/models/donation_spec.rb [hunk 2]
```diff
@@ -70,9 +69,11 @@
     end
   end
 
-  def stub_person_find_or_initialize_by
-    person = spy('person')
-    allow(Person).to receive(:find_or_initialize_by).and_return(person)
+  def stub_person
```
Comment (turino) [inline spec/models/donation_spec.rb]: Can you use `build_stubbed(:person)` instead of this method?

**Comment by turino** [comment]
I was hoping to keep person creation in the donation form object. We started talking about it in this pull request: https://github.com/MayOneUS/mayday-2.0-backend/pull/72, but I guess we never really finished the conversation. A couple reasons I may not have mentioned in that discussion:
 - I would like to be more explicit about what parameters we're expecting (some person params may not be appropriate here)
 - creating the person in the controller results in an extra call to update nationbuilder
 - one of the purposes of the form object was to avoid creating multiple objects in the controller (keeps controller code very simple)

I think I understand why you want to do it this way, though. Notes on these changes:
 - `occupation` and `employer` are basically remote fields on person -- should they be in `person_params` or `donation_params`?
 -  you need to ensure that person is valid and has email (person could technically be valid without email)
 - donation create action should return person errors in addition to donation errors

File: app/controllers/v1/donations_controller.rb (hunk 1 above)
Comment (Rio517) [inline app/controllers/v1/donations_controller.rb]: Makes sense.  Agreed.

**Comment by Rio517** [comment]
Hey, 

Let's finish that conversation.  Quick response is:

* I like being explicit, but I think we'll need to be able to pass in a full person when creating donations.  We get a lot of info from third party payment providers that we need to pass in. 
* I really don't love the double calls to NB, I'd like to figure out a way to avoid that, but don't have bandwidth to go through that right now. 
* I don't agree about not creating multiple objects here.  I think that it is good to minimize, but I don't want to pass in every single person param and then have our donation instance responsible for creating it.

secondary points:
* yeah, occupation and employer should be on person.  I didn't work through all the control flow logic, but would prefer to remove the redundancy and just have them on person. 
* Forgive me, but can you remind me why we must have an email on person in this case? I don't think nationbuilder or stripe needs that in some cases.
* Ahh good thought on validation errors.. Did I overlook this?  I'll have to take a closer look here.

File: app/controllers/v1/base_controller.rb [hunk 3]
```diff
@@ -11,17 +14,30 @@ def rescue_error_messages
...
+
+  def record_not_found(error)
+    render json: {error: error.message}, status: :not_found
+  end
+
+  def default_person_params
+    params.require(:person).permit(Person::ALL_AVAILABLE_ATTRIBUTES)
+  end
+
+  def person_from_params
```
Comment (turino) [inline app/controllers/v1/base_controller.rb]: hmmm... I see the benefit to this, as far as DRYing up the code goes, but I'm not a fan of using BaseController as grab bag of publicly available methods. Makes it harder to parse what's happening on the controllers that call these methods, and BaseController could end up getting pretty messy

File: app/models/donation.rb [hunk 4]
```diff
@@ -43,7 +40,7 @@ def process_payment
   end
 
   def record_donation
-    person = { email: email, employer: employer, occupation: occupation }
+    person = { email: self.person.email, employer: employer, occupation: occupation }
```
Comment (turino) [inline app/models/donation.rb]: could rename `person` local variable and not need to call self in `self.person.email`

File: app/models/donation.rb [hunk 5]
```diff
@@ -76,7 +69,7 @@ def create_stripe_charge
     Stripe::Charge.create(amount: amount_as_integer,
                           source: stripe_token,
                           currency: 'usd',
-                          description: "donation from #{email}")
+                          description: "donation from #{self.person.email}")
```
Comment (turino) [inline app/models/donation.rb]: why `self`?

File: app/models/person.rb [hunk 6]
```diff
@@ -52,11 +52,12 @@ class Person < ActiveRecord::Base
   delegate :update_location, :district, :state, to: :location
 
   FIELDS_ALSO_ON_NB = %w[email first_name last_name is_volunteer phone]
-  PERMITTED_PUBLIC_FIELDS = [:email, :phone, :first_name, :last_name, :address, :city, :state_abbrev, :zip, :is_volunteer, remote_fields: [:event_id, :employer, :occupation, :skills, tags: []]]
+  PERMITTED_PUBLIC_FIELDS = [:email, :phone, :first_name, :last_name, :full_name, :address, :city, :state_abbrev, :zip, :is_volunteer, remote_fields: [:event_id, :employer, :occupation, :skills, tags: []]]
```
Comment (turino) [inline app/models/person.rb]: what's `full_name`? Is that an NB property?

**Comment by turino** [comment]
Re: requiring email, right now our stripe methods call `person.email`. This is necessary for subscriptions. Could be changed for single charges (it's just being used in the charge description). I think it's just easier to say that a donation requires an email rather than say it requires email if charge is recurring, but you can do it either way. 

Re: validation, as it's written now, `person_from_params` is calling `Person.create_or_update`, which may or may not return a persisted record. I think you need to check that the returned person is valid/persisted. (I would still prefer to avoid using `person_from_params` method on `BaseController`, though)

File: app/controllers/v1/base_controller.rb (hunk 3 above)
Comment (turino) [inline app/controllers/v1/base_controller.rb]: can use `PersonConstructor`, instead: https://github.com/MayOneUS/mayday-2.0-backend/commit/3665af6e0b6e4eba55780312ff392e5db48ac8a1

but I guess I need to finish that...

**Comment by turino** [comment]
btw, I haven't noticed any change in spec run time

File: app/controllers/v1/base_controller.rb (hunk 3 above)
Comment (Rio517) [inline app/controllers/v1/base_controller.rb]: I agree for the params fields.  I'm sorta assuming we'll pull these peron params out when we have our constructor setup.  For the error handling, I'm not sure there is a better place.

File: app/models/person.rb (hunk 6 above)
Comment (Rio517) [inline app/models/person.rb]: Yes.  Exactly.

**Comment by Rio517** [comment]
For now I'm going to merge this for the hackathon this weekend, but will  create a new issue re: validation checking and cleaning up duplicate methods

File: app/models/person.rb (hunk 6 above)
Comment (turino) [inline app/models/person.rb]: As far as I can tell, `:full_name` can be passed in, but nothing will happen with it if it does. Are we supposed to try to split it and save it locally as first and last or just pass it directly to NB as is? Is there a form on the frontend that passes in full_name?

Comment (Rio517) [inline app/models/person.rb]: Hi,

Yes, it is handled on the nationbuilder side.  I was planning on using 
it for the new donation page, but might reverse this decision.

-Mario

**Pull request merged**